File: ch1/js/ka-stream.js
import { ChatAnthropic } from '@langchain/anthropic';
import { ChatPromptTemplate } from '@langchain/core/prompts';
import { RunnableLambda } from '@langchain/core/runnables';

const template = ChatPromptTemplate.fromMessages([
  ['system', 'You are a helpful assistant.'],
  ['human', '{question}'],
]);

const model = new ChatAnthropic({
  model: 'claude-haiku-4-5',
});

const chatbot = RunnableLambda.from(async function* (values) {
  const prompt = await template.invoke(values);
  for await (const token of await model.stream(prompt)) {
    yield token;
  }
});

for await (const token of await chatbot.stream({
  question: 'Which model providers offer LLMs?',
})) {
  console.log(token);
}

File: ch1/py/ka-stream.py
from langchain_core.runnables import chain
from langchain_anthropic import ChatAnthropic
from langchain_core.prompts import ChatPromptTemplate


model = ChatAnthropic(model="claude-haiku-4-5")


template = ChatPromptTemplate.from_messages(
    [
        ("system", "You are a helpful assistant."),
        ("human", "{question}"),
    ]
)


@chain
def chatbot(values):
    prompt = template.invoke(values)
    for token in model.stream(prompt):
        yield token


for part in chatbot.stream({"question": "Which model providers offer LLMs?"}):
    print(part)
